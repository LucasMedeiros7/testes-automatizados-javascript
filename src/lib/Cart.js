import Dinero from 'dinero.js';

const Money = Dinero;
Money.defaultCurrency = 'BRL';
Money.defaultPrecision = 2;

export class Cart {
  items = [];

  add(item) {
    const productIndex = this.items.findIndex(
      ({ product }) => product === item.product
    );
    if (productIndex != -1) this.items.splice(productIndex, 1);
    this.items.push(item);
  }

  getTotal() {
    const total = this.items.reduce((acc, item) => {
      return acc.add(Money({ amount: item.quantity * item.product.price }));
    }, Money({ amount: 0 }));
    return total;
  }

  remove(product) {
    const productIndex = this.items.findIndex(item => item.product === product);
    this.items.splice(productIndex, 1);
  }

  summary() {
    const total = this.getTotal().getAmount();
    const items = this.items;
    return {
      total,
      items,
    };
  }

  checkout() {
    const { total, items } = this.summary();
    this.items = [];
    return {
      total,
      items,
    };
  }
}
