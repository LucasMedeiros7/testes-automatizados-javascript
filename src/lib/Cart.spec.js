import { Cart } from "./Cart.js";

describe("Cart", () => {
  let cart;
  let product = {
    title: "NIKE SB DUNK LOW PRO",
    price: 65388, // 353.88 | R$ 383,88
  };

  beforeEach(() => {
    cart = new Cart();
  });

  it("should return 0 when getTotal() is executed in a newly created instance", () => {
    expect(cart.getTotal()).toBe(0);
  });

  it("should multiply quantity and price and receive the total amount", () => {
    const item = {
      product,
      quantity: 2, // 130776
    };

    cart.add(item);

    expect(cart.getTotal()).toEqual(130776);
  });

  it("should ensure no more than on product exists at a time", () => {
    cart.add({
      product,
      quantity: 2,
    });

    cart.add({
      product,
      quantity: 1,
    });

    expect(cart.getTotal()).toEqual(65388);
  });
});
