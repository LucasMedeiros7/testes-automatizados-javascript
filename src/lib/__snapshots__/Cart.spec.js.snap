// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart checkout() should return an object with the total and the list of items 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 65388,
        "title": "NIKE SB DUNK LOW PRO",
      },
      "quantity": 5,
    },
    Object {
      "product": Object {
        "price": 135388,
        "title": "JORDAN 1 LOW",
      },
      "quantity": 3,
    },
  ],
  "total": 733104,
}
`;

exports[`Cart checkout() should return an object with the total and the list of items when summary() is called 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 65388,
        "title": "NIKE SB DUNK LOW PRO",
      },
      "quantity": 5,
    },
    Object {
      "product": Object {
        "price": 135388,
        "title": "JORDAN 1 LOW",
      },
      "quantity": 3,
    },
  ],
  "total": 733104,
}
`;
